File: 836-rectangle-overlap/solution.py
class Solution:
    def isRectangleOverlap(self, rec1: list[int], rec2: list[int]) -> bool:
        # rec = [x1, y1, x2, y2]
        # Check if the 1D projections overlap on both the X-axis and Y-axis
        x_overlap = rec1[0] < rec2[2] and rec2[0] < rec1[2]
        y_overlap = rec1[1] < rec2[3] and rec2[1] < rec1[3]
        
        return x_overlap and y_overlap

File: 836-rectangle-overlap/solution.cpp
#include <vector>

class Solution {
public:
    bool isRectangleOverlap(const std::vector<int>& rec1, const std::vector<int>& rec2) {
        // rec = [x1, y1, x2, y2]
        // Check if horizontal projections overlap:
        // rec1's left < rec2's right AND rec2's left < rec1's right
        bool xOverlap = rec1[0] < rec2[2] && rec2[0] < rec1[2];
        
        // Check if vertical projections overlap:
        // rec1's bottom < rec2's top AND rec2's bottom < rec1's top
        bool yOverlap = rec1[1] < rec2[3] && rec2[1] < rec1[3];
        
        return xOverlap && yOverlap;
    }
};